src/WorldState/WorldStateClient.cs

```diff
@@ -95,12 +98,120 @@ public Task<ConclaveChallenge[]> ConclaveChallenges()
             return Get<ConclaveChallenge[]>(builder);
         }
 
+        public Task<DailyDeal[]> DailyDeals()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.DailyDeal;
+            return Get<DailyDeal[]>(builder);
+        }
+
+        public Task<EarthCycle> EarthCycle()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.EarthRotation;
+            return Get<EarthCycle>(builder);
+        }
+
+        public Task<Event[]> Events()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.Events;
+            return Get<Event[]>(builder);
+        }
+
+        public Task<FlashSale[]> FlashSales()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.FlashSales;
+            return Get<FlashSale[]>(builder);
+        }
+
         public Task<ConstructionProgress> FleetConstructionProgress()
         {
             builder.Path = platformAsString + WorldStateEndpoints.Construction;
             return Get<ConstructionProgress>(builder);
         }
 
+        public Task<GlobalUpgrade[]> GlobalUpgrades()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.GlobalBoosts;
+            return Get<GlobalUpgrade[]>(builder);
+        }
+
+        public Task<Invasion[]> Invasions()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.Invasions;
+            return Get<Invasion[]>(builder);
+        }
+
+        public Task<News[]> News()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.News;
+            return Get<News[]>(builder);
+        }
+
+        public Task<Nightwave> Nightwave()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.Nightwave;
+            return Get<Nightwave>(builder);
+        }
+
+        public Task<OrbVallisCycle> OrbVallisCycle()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.OrbVallis;
+            return Get<OrbVallisCycle>(builder);
+        }
+
+        public async Task<RivenMod[]> RivenMods()
+        {
+            builder.Path = platformAsString + WorldStateEndpoints.Rivens;
+            var json = (await Get<JToken>(builder).ConfigureAwait(false));
+
+            if (json == null) return Array.Empty<RivenMod>();
+
+            var mods = new List<RivenMod>(350);
+            foreach (var category in json.Values())
+            {
+                foreach (var weapon in category.Values())
+                {
+                    foreach (var variant in weapon.Values())
+                    {
+                        if (!variant.HasValues) continue;
+
+                        mods.Add(variant.ToObject<RivenMod>());
+                    }
+                }
+            }
+
+            return mods.ToArray();
+        }
```

Comment:
Riven needs more work performance-wise. Those `foreach`es could've be done during deserialization, so the entire JSON (as a `JToken`) never needed to be read in memory.

For further optimization, all collection endpoints (anything that returns an array) could be made async-enumerated when targeting .NET Standard 2.1 (in an extension project?).

Reply:
are you fetching each endpoint individually when this parses all of them?

Reply:
are you intending to fetch each endpoint individually (aka not the whole platform) for the dynamic data endpoints whenever this fetches?

Reply:
Currently, yes. I think breaking functions apart is the most common use case. At least for me, for example I sometimes just check Baro’s inventory and couldn’t care less about what’s enemy’s fleet construction progress or what’s the active challenge. Fetching the entire World State puts too much strain on the network for that.

But since you said that I will add an aggregated model and corresponding method to pull the whole thing.

Reply:
i don't think i'd mind if it was, like, an option to grab one individually if a specific one was set to auto-update on it's own, but if you're updating the whole thing and each concept is grabbing it's endpoint (every time), that'd get a bit troublesome.

i mean, they'd still probably be cached copies, so it shouldn't be a huge deal, but at the same time, i'd try to minimize hits...

Reply:
Don’t worry. I’ll definitely pull the endpoint that has the whole package if I need the entire world state.

Right now this client looks like more “utility” than a “client”. Ideally I imagine `WorldStateClient` would be a long-lived (in-memory) mirror of the entire world state, with internal timer/scheduler that keeps it up-to-date, as long as it’s not disposed. When that happens, I’ll then move all these one-time getters to probably a static class.

Anyway, I’m still working on minimizing the memory footprint of JSON. Because I think this is a priority and I can learn a great deal from these experiments.

Reply:
Also, on the topic of the first comment in this thread, System.Interactive has `System.IAsyncEnumerable` as part of its library. If you really wanted to, you could take advantage of it, but the project would need to have netstandard2.1 as an additional target to make sure that type isn't included if the project using this library targets that version of the standard. Regardless, I don't think it really would benefit this method. This method requests all the rivens at once and receives that request as a single bulk request. `IAsyncEnumerable` is better if you have to deal with pagination since async enumeration would allow you to make more requests as needed later on. In this case, `Task<IEnumerable<RivenMod>>`, `Task<RivenMod[]>`, or whatever works perfectly fine.

Reply:
On the topic of AsyncEnumerable, I'm working on something that breaks the JSON response apart (i.e., not downloads it first in its entirety and then deserializes, unlike traditional API calls) and streams/enumerates one model at a time for efficiency and bandwidth concerns. It's probably gonna go into another project (or class) so consumers can have options. I understand it sounds unconventional but I myself do have a special usecase for that.

First thing first I do plan on making all collections `IEnumerable<>` at least in this project.

Reply:
How will you break the request apart if [the server sends all the riven data at once](https://api.warframestat.us/pc/rivens)?

Reply:
Since we're accepting a network stream, the stream could be paired up with a `JsonReader` and manually read, partially deserialized without calling `JsonConvert` on the entire JSON, which is basically what's happening right now although not as tweaked.

Say you want the first riven (or alerts), read/enumerate 1, continue or close the request/stream.

I tried this on the _Weapons_ endpoint, which returns a JSON of about 7 MB. It only benefits from use cases where you don't need all of them at once, e.g. populating a table asyncly. And the advantage/disadvantage (depends) is that the connection may drop prematurely and only partial data is collected, in contrast to traditionally all-or-nothing API calls.